param_sync.py

```diff
@@ -7,12 +7,39 @@ class ParamSyncRule:
     training.
     """
 
+    def update_core(self, local_params, master_params):
+        """
+        Iterator over updates for local and master params.
+
+        This function must be ready to deal with symbolic or numerical
+        inputs.
+        """
+        raise NotImplementedError()
+
     def update_params(self, local_params, master_params):
         """
         Perform an inplace update of the master parameters and returns the
         parameters based on a certain parameter synchronisation rule.
         """
-        raise NotImplementedError()
+        for lp, mp, ul, um in izip(local_params, master_params,
+                                   self.update_core(local_params,
+                                                    master_params)):
+            lp += ul
+            mp += um
+
+    def update_fn(self, local_params):
+        """
+        Return a theano function that will update the local shared
+        params and return the new values for the master params.
+        """
+        from theano import function
```

Comment:
I am under the (possibly mistaken) impression that this *can* work without theano and that we should try to keep it that way without burdening the code too much.

If I'm wrong and we can assume theano, then I'll revise the code.